### ISSUE TITLE
Use different default sell prices for patient dispensing vs outbound shipments

### ISSUE THREAD

**Issue by adamdewey** [msupply-foundation/open-msupply#4703]
## Is your feature request related to a problem? Please describe 👀

A requested use case is for one default sell price to be used for when shipping stock to downstream facilities and a different default sell price to be used when dispensing to patients. 

Some Stores may do both - e.g. a hospital sending stock to smaller pharmacies and also dispensing directly to patients. 

## Describe the solution you'd like 🎁

Here's the pricing workflow:

![image](https://github.com/user-attachments/assets/b9d79f4e-7645-441f-ad40-df1bab24ff13)

## Configuration
( All are items are changes to make in mSupply )
* Create a master list for all items, check the `is_default_price_list` option
* This `is_default_price_list` applies to all stores, and doesn't have to be assigned to them...
* Define prices for each item 
* Create a master list for all discounted items, set the `discount` percentage value
* This discount master list need to be assigned to any stores that should receive the discount rate.
 
## Changes required

* update master lists
  * add `is_default_price_list` to `master_list`
  * add `discount` to `master_list` 
  * and update sync and translation layers
  * add `price` to `master_list_line` and update sync & translation layers

* When stock is issued apply the `is_default_price_list` prices to the invoice line (as appropriate)
* Apply discount to outgoing invoices (apart from dispensing)

**Stocktake**
_No changes_

...OUT OF SCOPE -  POSSIBLE FUTURE REQUIREMENTS...
```
- [ ] If new stock line is introduced by stocktake, it's sell price should be set to price from `is_default_price_list`
- [ ] Stocklines are introduced from either Add Item, or Edit Item (if the stock take is created from a master list) (FRONTEND)
- [ ] Cost prices are assumed to be discounted if discount would apply to this store
- [ ] Sell Prices are not editable by the user 
```

**View Stock**
_No changes_

...OUT OF SCOPE - POSSIBLE FUTURE REQUIREMENTS...
```
- [ ] Sell Price should be prepopulated based on `is_default_price_list` once a Item is selected in the New Stock Modal
![image](https://github.com/user-attachments/assets/a5ba6642-cca7-4699-a647-5799c6168471)
- [ ] Cost prices are assumed to be discounted if discount would apply to this store
- [ ] Sell Prices are not editable by the user 
```

**Outbound Shipment**
- [x] Sell price should looked up from the `is_default_price_list`
- [x] However, a __discount__ should be applied based on the Customer's(Name) discount master list
   - [x] The global discount should apply from the masterlist to all items that are in that discount master list
   - [x] If an item is not in the discount master list no discount should be applied
- [x] Add Item (Frontend) => Modal runs graphql query to get Discount & apply to `invoice_line`)
- [x] Actions-> Allocate Lines (Backend: `allocatedOutboundShipmentUnallocatedLines`)  => Backend to look up default price list and apply price + any discount automatically
- [x] Add from Master List (Backend:`addToOutboundShipmentFromMasterList` ) => _No Changes required as pricing will be applied at allocation time_

**Inbound Shipment**
_No changes_

...OUT OF SCOPE -  POSSIBLE FUTURE REQUIREMENTS...
```
- [ ] Sell price should looked up from the `is_default_price_list`
- [ ] Discount will be applied based on the store name's discount list whether it comes via a transfer or otherwise
- [ ] Add from Master List (`addToInboundShipmentFromMasterList`) => look up `is_default_price_list` price for that item and apply it to `sell_price` 
- [ ] Cost price should come from sell price on inbound shipment
- [ ] Add Item  Frontend) => Modal runs graphql query to get`is_default_price_list` and applies it to `Sell Price`
- [ ] Cost prices are assumed to be discounted if discount would apply to this store
- [ ] Sell Prices are not editable by the user 
```

**Internal Orders**
_No changes required_

**Requsitions, R&R**
_No changes required_

**Returns**
_No changes required_

**Transfer aka in Processors code**
_No changes required_

**Dispensing**
- [x] Sell price should looked up from the `is_default_price_list` (Stockline Price is ignored)
- [x] No discounts are applied during dispensing

### Additional context 💌

Related to https://github.com/msupply-foundation/msupply/issues/15108

### Moneyworks Jobcode 🧰

`COMSCUST`

**Comment by andreievg** [comment]
Places where discount should apply:

Outbound Shipment "Add Item" - Front End ([example](https://github.com/msupply-foundation/open-msupply/commit/63f5643319bf97fcaf22c3cf9f59073d65121af7))

Outbound Shipment "Allocate Placeholder Line" (line select) - Back End

The places where stock_line (and associated invoice_line) is created:

Inbound Shipment "Add Item" - Front End

Inbound Shipment "Add From Master List" - Back End

Stock View "New Stock" - Front End

Stocktake "Add From Master List" - Back End

Stocktake "Add Item", "Add Batch" - Front End

**Comment by lache-melvin** [comment]
@jmbrunskill is this issue closed now?

**Comment by jmbrunskill** [comment]
Yes, closing, sorry!

**Comment by jmbrunskill** [comment]
@lache-melvin Actually, we need to merge the v2.2.2 branch into main...

**Comment by lache-melvin** [comment]
Is it merged into develop yet @jmbrunskill ?

**Comment by jmbrunskill** [comment]
Nope not merged anywhere, was waiting on RC testing to be completed first...

**Comment by AnushaUp** [comment]
# 🧪 Testing
Tested in  OMS `V2.2.02-RC1` and mSupply `V7.17.08`


- [ ] Update mSupply to include a PR for this ISSUE -> https://github.com/msupply-foundation/msupply/issues/15108 (doesn't exist yet)
- [ ] Add a master list with is_default_price_list set
- [ ] Add a masterlist with a discount set
- [ ] Sync open-mSupply
- [ ] Check the fields are set correctly....
- [ ] Check legacy mSupply master lists still sync correctly as well

**Comment by AnushaUp** [comment]
Passed from OMS side👍✅
